pallets/sponsorships/src/sponsored.rs

```diff
@@ -0,0 +1,205 @@
+use codec::{Decode, Encode};
+use frame_support::{
+	dispatch::{DispatchInfo, DispatchResult, PostDispatchInfo},
+	traits::IsSubType,
+};
+use pallet_revive::AddressMapper;
+use pallet_transaction_payment::OnChargeTransaction;
+use scale_info::{StaticTypeInfo, TypeInfo};
+use sp_core::U256;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, PostDispatchInfoOf, SignedExtension},
+	transaction_validity::{TransactionValidity, TransactionValidityError},
+};
+use types::*;
+
+use super::*;
+
+/// A [`SignedExtension`] that handles fees sponsorship.
+#[derive(Encode, Decode, Clone, Eq, PartialEq)]
+pub struct Sponsored<T, S>(pub S, core::marker::PhantomData<T>);
+
+// Make this extension "invisible" from the outside (ie metadata type information)
+impl<T, S: StaticTypeInfo> TypeInfo for Sponsored<T, S> {
+	type Identity = S;
+
+	fn type_info() -> scale_info::Type {
+		S::type_info()
+	}
+}
+
+impl<T, S: Encode> core::fmt::Debug for Sponsored<T, S> {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+		write!(f, "Sponsored<{:?}>", self.0.encode())
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut core::fmt::Formatter) -> core::fmt::Result {
+		Ok(())
+	}
+}
+#[cfg(test)]
+impl<T, S> From<S> for Sponsored<T, S> {
+	fn from(s: S) -> Self {
+		Self(s, core::marker::PhantomData)
+	}
+}
+
+impl<T: Config, S> Sponsored<T, S>
+where
+	<T as frame_system::Config>::RuntimeCall: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>
+		+ IsSubType<Call<T>>
+		+ IsSubType<pallet_revive::Call<T>>,
+	<<T as pallet_revive::Config>::Currency as frame_support::traits::fungible::Inspect<
+		<T as frame_system::Config>::AccountId,
+	>>::Balance: Into<U256>,
+	<<T as pallet_revive::Config>::Currency as frame_support::traits::fungible::Inspect<
+		<T as frame_system::Config>::AccountId,
+	>>::Balance: TryFrom<U256>,
+	<<T as pallet_revive::Config>::Time as frame_support::traits::Time>::Moment: Into<U256>,
+{
+	fn is_contracts_call(
+		call: &<T as frame_system::Config>::RuntimeCall,
+	) -> Option<AccountIdOf<T>> {
+		match call.is_sub_type() {
+			Some(pallet_revive::Call::<T>::call { dest, .. }) => {
```

Comment:
Original comment from https://github.com/r0gue-io/pop-node/pull/351#discussion_r1816359073



Ideally this would be passed in via a generic type parameter which resolves a runtime call to an account id so no tight coupling to revive and configured at runtime level.

No changes required.